atlp-rwanda/e-commerce-lydia-32-bn
# Ft 2fa 187570869: Two Two-factor authentication for Sellers

### What does this PR do?
This Pull Request (PR) introduces a two-factor authentication (2FA) mechanism for sellers in the application. It allows sellers to receive a one-time code via email during the login process, which they must provide to complete the authentication process successfully.

### Description of Task to be completed?
The main tasks completed in this PR are:

- Generate 2FA Code: During the login process, if the user is identified as a seller, a random 5-digit code is generated and stored in the twoFactorSecret column of the user's record.

- Send 2FA Code via Email: The generated 2FA code is sent to the seller's email address using a custom email sending functionality (e.g., nodemailer).

- Verify 2FA Code: A separate endpoint is created to verify the 2FA code provided by the seller. If the provided code matches the stored twoFactorSecret, the twoFactorSecret is cleared from the user's record, and a JWT token is generated and sent back as a cookie.

- Handle Unsuccessful Verification: If the provided 2FA code does not match the stored code, an error response is sent back to the client.

### How should this be manually tested?
To manually test the 2FA functionality, follow these steps:

1. Log in as a seller using valid credentials.
2.  After successful authentication, check your email inbox for the 2FA code.
3. Send a POST request to the /verify-two-factor endpoint with the following payload:

{
  "twoFactorCode": <the_2fa_code_received_via_email>
}
4. If the 2FA code is valid, you should receive a response with a JWT token in a cookie.
5. Verify that the twoFactorSecret column in the user's record is now cleared.
6. Try sending an invalid 2FA code to the /factor endpoint and ensure you receive an error response.

### Any background context you want to provide?
Two-factor authentication adds an extra layer of security to the login process, especially for privileged user roles like sellers. By requiring a one-time code in addition to the password, it helps mitigate the risk of unauthorized access even if the password is compromised.

### What are the relevant pivotal tracker stories
187570869
